File: leetcode/algorithms/c++/Clone Graph/Clone Graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution
{
private:
    map<Node *, Node *> v;

public:
    Node *cloneGraph(Node *node)
    {

        if (node == NULL)
            return NULL;
        if (v.find(node) != v.end())
            return v[node];

        Node *c = new Node(node->val);
        v[node] = c;
        for (auto x : node->neighbors)
        {
            auto i = cloneGraph(x);
            if (i != NULL)
                c->neighbors.push_back(i);
        }

        return c;
    }
};

File: leetcode/algorithms/c#/Clone Graph/Clone Graph.cs
/*
// Definition for a Node.
public class Node {
    public int val;
    public IList<Node> neighbors;

    public Node() {
        val = 0;
        neighbors = new List<Node>();
    }

    public Node(int _val) {
        val = _val;
        neighbors = new List<Node>();
    }

    public Node(int _val, List<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/
public class Solution
{
    private Dictionary<Node, Node> dict = new Dictionary<Node, Node>();

    public Node CloneGraph(Node node)
    {
        if (node == null) return null;
        if (dict.ContainsKey(node)) return dict[node];
        var res = new Node(node.val);
        dict.Add (node, res);
        foreach (var item in node.neighbors)
        res.neighbors.Add(CloneGraph(item));

        return res;
    }
}
